File: python/aps/compat.py
"""Compatibility bridges for Agent Skills, AGENTS.md."""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from .passport import Skill


def import_agent_skill(skill_dir: str) -> Skill:
    """Read an Agent Skills folder and convert to a Skill."""
    skill_md_path = Path(skill_dir) / "SKILL.md"
    content = skill_md_path.read_text(encoding="utf-8")
    name = Path(skill_dir).name
    description = _extract_first_line(content)
    capabilities = _infer_capabilities(content)
    return Skill(
        name=name,
        version="1.0.0",
        description=description,
        capabilities=capabilities,
        source=f"agent-skills://{name}",
        hash="",
    )


def export_agent_skill(skill: Skill, output_dir: str) -> None:
    """Write a Skill as an Agent Skills folder."""
    skill_dir = Path(output_dir) / skill.name
    skill_dir.mkdir(parents=True, exist_ok=True)

    md = f"# {skill.name}\n\n{skill.description}\n\nVersion: {skill.version}\n\n## Capabilities\n\n"
    for cap in skill.capabilities:
        md += f"- {cap}\n"
    (skill_dir / "SKILL.md").write_text(md, encoding="utf-8")

    meta = {"name": skill.name, "version": skill.version, "description": skill.description,
            "capabilities": skill.capabilities, "hash": skill.hash}
    if skill.source:
        meta["source"] = skill.source
    (skill_dir / "metadata.json").write_text(json.dumps(meta, indent=2), encoding="utf-8")


class AgentsMD:
    def __init__(self, raw: str = "", instructions: list[str] | None = None,
                 constraints: list[str] | None = None, tools: list[str] | None = None):
        self.raw = raw
        self.instructions = instructions or []
        self.constraints = constraints or []
        self.tools = tools or []


def load_agents_md(repo_path: str) -> AgentsMD:
    """Read and parse an AGENTS.md file."""
    agents_path = Path(repo_path) / "AGENTS.md"
    content = agents_path.read_text(encoding="utf-8")
    result = AgentsMD(raw=content)
    section = ""
    for line in content.split("\n"):
        trimmed = line.strip()
        lower = trimmed.lower()
        if trimmed.startswith("#"):
            if "instruction" in lower or "rules" in lower:
                section = "instructions"
            elif "constraint" in lower or "restriction" in lower:
                section = "constraints"
            elif "tool" in lower or "mcp" in lower:
                section = "tools"
            else:
                section = ""
            continue
        if trimmed.startswith("- ") or trimmed.startswith("* "):
            item = trimmed.lstrip("-* ")
            if section == "instructions":
                result.instructions.append(item)
            elif section == "constraints":
                result.constraints.append(item)
            elif section == "tools":
                result.tools.append(item)
    return result


def _extract_first_line(content: str) -> str:
    for line in content.split("\n"):
        trimmed = line.strip()
        if trimmed and not trimmed.startswith("#"):
            return trimmed
    return ""


def _infer_capabilities(content: str) -> list[str]:
    lower = content.lower()
    caps: list[str] = []
    if "code" in lower or "develop" in lower:
        caps.append("code_write")
    if "test" in lower:
        caps.append("test_run")
    if "debug" in lower:
        caps.append("debug")
    if "build" in lower:
        caps.append("build")
    if "review" in lower or "audit" in lower:
        caps.append("code_review")
    if "data" in lower or "analyz" in lower:
        caps.append("data_read")
    return caps or ["general"]

File: typescript/src/compat.ts
import { readFileSync, readdirSync, mkdirSync, writeFileSync } from 'fs';
import { join, basename } from 'path';
import type { Skill, AgentsMD } from './types.js';

export function importAgentSkill(skillDir: string): Skill {
  const content = readFileSync(join(skillDir, 'SKILL.md'), 'utf-8');
  const name = basename(skillDir);
  const description = extractFirstLine(content);
  const capabilities = inferCapabilities(content);

  return {
    name,
    version: '1.0.0',
    description,
    capabilities,
    source: `agent-skills://${name}`,
    hash: '',
  };
}

export function exportAgentSkill(skill: Skill, outputDir: string): void {
  const skillDir = join(outputDir, skill.name);
  mkdirSync(skillDir, { recursive: true });

  let md = `# ${skill.name}\n\n${skill.description}\n\nVersion: ${skill.version}\n\n## Capabilities\n\n`;
  for (const cap of skill.capabilities) {
    md += `- ${cap}\n`;
  }
  writeFileSync(join(skillDir, 'SKILL.md'), md);
  writeFileSync(join(skillDir, 'metadata.json'), JSON.stringify(skill, null, 2));
}

export function loadAgentsMd(repoPath: string): AgentsMD {
  const content = readFileSync(join(repoPath, 'AGENTS.md'), 'utf-8');
  const result: AgentsMD = { raw: content, instructions: [], constraints: [], tools: [] };

  let section = '';
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    const lower = trimmed.toLowerCase();

    if (trimmed.startsWith('#')) {
      if (lower.includes('instruction') || lower.includes('rules')) section = 'instructions';
      else if (lower.includes('constraint') || lower.includes('restriction')) section = 'constraints';
      else if (lower.includes('tool') || lower.includes('mcp')) section = 'tools';
      else section = '';
      continue;
    }

    if (trimmed.startsWith('- ') || trimmed.startsWith('* ')) {
      const item = trimmed.replace(/^[-*]\s+/, '');
      if (section === 'instructions') result.instructions.push(item);
      else if (section === 'constraints') result.constraints.push(item);
      else if (section === 'tools') result.tools.push(item);
    }
  }

  return result;
}

function extractFirstLine(content: string): string {
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#')) return trimmed;
  }
  return '';
}

function inferCapabilities(content: string): string[] {
  const lower = content.toLowerCase();
  const caps: string[] = [];
  if (lower.includes('code') || lower.includes('develop')) caps.push('code_write');
  if (lower.includes('test')) caps.push('test_run');
  if (lower.includes('debug')) caps.push('debug');
  if (lower.includes('build')) caps.push('build');
  if (lower.includes('review') || lower.includes('audit')) caps.push('code_review');
  if (lower.includes('data') || lower.includes('analyz')) caps.push('data_read');
  if (caps.length === 0) caps.push('general');
  return caps;
}
